### PR TITLE
Iontransfer

### PR DESCRIPTION
**Pull request by nrnhines**
**Extend CoreNEURON gap transfer to be equivalent to NEURON**
Associated with neuronsimulator/nrn#883

- [x] Issue neuronsimulator/nrn#842 fixed
- [x] Issue neuronsimulator/nrn#866 fixed
- [x] Generalized to non-voltage sources.

**How to test this?**

```bash
cmake .. -DCMAKE_INSTALL_PREFIX=install -DIV_DIR=$HOME/neuron/ivcmake/build/install -DPYTHON_EXECUTABLE=`which python3` -DNRN_ENABLE_PYTHON_DYNAMIC=ON -DNRN_PYTHON_DYNAMIC="python2.7;python3.7" -DNRN_ENABLE_TESTS=ON -DNRN_ENABLE_CORENEURON=ON -DCMAKE_BUILD_TYPE=Debug
make -j 6 install
cd ../test/gjtests
nrnivmodl -coreneuron mod
mpiexec -n 2 nrniv -mpi -python test_natrans.py
```
File transfer mode not tested.

CI_BRANCHES:NEURON_BRANCH=iontransfer,

### CODE REVIEW

**Comment by pramodk** [comment]
>  Generalized to non-voltage sources.

@nrnhines : I see above item is not checked yet. Is that still work in progress? Or should I review this already?

Retest this please.

**Comment by nrnhines** [comment]
I believe this is ready to merge. 
```
valgrind --leak-check=full --show-leak-kinds=definite `pyenv which python` test_natrans.py
```
does not show any definite leaks that were introduced by this pull request. Do all the unsuccessful checks just mean that the data files are 1.4 and neuron master is still 1.3? Or that all the test data files need to be rebuilt?

**Comment by pramodk** [comment]
> Do all the unsuccessful checks just mean that the data files are 1.4 and neuron master is still 1.3? Or that all the test data files need to be rebuilt?

yes, we need to regenerate test data:

>  num_mpi=2
> num_omp_thread=1
>
> Error: Incompatible binary input dataset version (expected 1.4, input 1.3)
> MPT ERROR: Rank 1(g:1) received signal SIGABRT/SIGIOT(6).

We have instructions to generate data here: https://github.com/BlueBrain/CoreNeuron/tree/master/tests/integration

I will take a look at other CI tests tomorrow.

**Comment by nrnhines** [comment]
It appears that there are other tests that fail with  ```Error: Incompatible binary input dataset version (expected 1.4, input 1.3)``` in the Simulation Stack build. I'm not sure how to interpret the  errors in the Jenkins pipeline for the 
```
Failed in branch GPU-unified
Failed in branch GPU-non-unified
...
100% tests passed, 0 tests failed out of 15
```
It is certainly the case that I did not test this pull request on a gpu.

**Comment by pramodk** [comment]
You can see log here: https://bbpcode.epfl.ch/ci/blue/organizations/jenkins/oss.coreneuron.pipeline/detail/oss.coreneuron.pipeline/1618/pipeline

```

[ 69%] Building CXX object coreneuron/CMakeFiles/coreneuron.dir/network/partrans_setup.cpp.o
cd /jenkins/07/workspace/oss.coreneuron.pipeline/build_GPU-non-unified/coreneuron && /gpfs/bbp.cscs.ch/ssd/apps/hpc/jenkins/deploy/compilers/2020-02-01/linux-rhel7-x86_64/gcc-4.8.5/pgi-19.10-7s7noo24zw/linux86-64-llvm/19.10/bin/pgc++  -DCORENEURON_BUILD -DCUDA_PROFILING -DDISABLE_HOC_EXP -DEIGEN_DONT_VECTORIZE=1 -DENABLE_SPLAYTREE_QUEUING -DHAVE_MALLOC_H -DLAYOUT=0 -DMPICH_SKIP_MPICXX=1 -DMPI_NO_CPPBIND=1 -DNRNMPI=1 -DOMPI_SKIP_MPICXX=1 -DPG_ACC_BUGS -DSWAP_ENDIAN_DISABLE_ASM -I/jenkins/07/workspace/oss.coreneuron.pipeline/coreneuron/utils/randoms -I/gpfs/bbp.cscs.ch/ssd/apps/hpc/jenkins/deploy/tools/2020-02-01/linux-rhel7-x86_64/gcc-8.3.0/hpe-mpi-2.21-7pbszh6v5u/include -I/jenkins/07/workspace/oss.coreneuron.pipeline/build_GPU-non-unified/coreneuron -I/jenkins/07/workspace/oss.coreneuron.pipeline/build_GPU-non-unified/include -I/gpfs/bbp.cscs.ch/ssd/apps/hpc/jenkins/deploy/external-libraries/2020-02-01/linux-rhel7-x86_64/gcc-8.3.0/cuda-9.2.88-vwflnw2g4n/include -I/jenkins/07/workspace/oss.coreneuron.pipeline -I/jenkins/07/workspace/oss.coreneuron.pipeline/external/Random123/include -I/jenkins/07/workspace/oss.coreneuron.pipeline/external/CLI11/include  -acc  --diag_suppress 161,177,550 -mp  -g  -O2 -fPIC   --c++14 -o CMakeFiles/coreneuron.dir/network/partrans_setup.cpp.o -c /jenkins/07/workspace/oss.coreneuron.pipeline/coreneuron/network/partrans_setup.cpp


"/jenkins/07/workspace/oss.coreneuron.pipeline/coreneuron/network/partrans.cpp"
          , line 110: error: identifier "nt" is undefined
          tar_data[0:nt._ndata],                      \
                     ^

"/jenkins/07/workspace/oss.coreneuron.pipeline/coreneuron/network/partrans.cpp"
          , line 135: error: expected a "]"
              #pragma acc enter data copyin(ttd.src_indices.data()[0 : n]) if (nrn_threads[0].compute_gpu)
                                                                     ^

"/jenkins/07/workspace/oss.coreneuron.pipeline/coreneuron/network/partrans.cpp"
          , line 136: error: a pointer to a bound function may only be used to
          call the function
              #pragma acc enter data create(ttd.src_gather.data[0 : ng]) if (nrn_threads[0].compute_gpu)
                                                           ^

"/jenkins/07/workspace/oss.coreneuron.pipeline/coreneuron/network/partrans.cpp"
          , line 142: error: expected a "]"
              #pragma acc enter data copyin(ttd.insrc_indices.data()[0 : ttd.insrc_indices.size()]) if (nrn_threads[0].compute_gpu)

```

I will have time later this evening and then I can test/fix this.

**Comment by pramodk** [comment]
For simulation stack, the pipeline launched here was successful : https://bbpcode.epfl.ch/ci/blue/organizations/jenkins/hpc.SimulationStack/detail/hpc.SimulationStack/2885/pipeline/

File: coreneuron/io/nrn_setup.cpp [hunk 1]
```diff
@@ -528,6 +512,28 @@ void nrn_setup(const char* filesdat,
     /* nrn_multithread_job supports serial, pthread, and openmp. */
     coreneuron::phase_wrapper<coreneuron::phase::two>(userParams, corenrn_embedded);
 
+    // gap junctions
+    // Gaps are done after phase2, in order to use layout and permutation
+    // information via calls to stdindex2ptr.
+    if (nrn_have_gaps) {
+        nrn_partrans::transfer_thread_data_ = new nrn_partrans::TransferThreadData[nrn_nthread];
```
Comment (pramodk) [inline coreneuron/io/nrn_setup.cpp:519]: @nrnhines : I couldn't see where this `transfer_thread_data_ ` is deallocated. could you check?

**Review by pramodk — APPROVED** [review]
LGTM.

CI is failing because NEURON_BRANCH=iontransfer is not working. I have manually tested CI with Spack branch and everything is fine.

**Pull request merged**